arangod/Aql/Query.cpp

```diff
@@ -266,20 +266,20 @@ void Query::setLockTimeout(double timeout) noexcept {
 }
 
 bool Query::killed() const {
-  if (_queryOptions.maxRuntime > std::numeric_limits<double>::epsilon() &&
-      elapsedSince(_startTime) > _queryOptions.maxRuntime) {
-    return true;
-  }
-  return _queryKilled;
+  return (std::numeric_limits<double>::epsilon() < _queryOptions.maxRuntime &&
+          _queryOptions.maxRuntime < elapsedSince(_startTime)) ||
+         _queryKilled.load(std::memory_order_acquire);
 }
 
 /// @brief set the query to killed
 void Query::kill() {
-  if (ServerState::instance()->isCoordinator() && !_queryKilled) {
-    _queryKilled = true;
-    this->cleanupPlanAndEngine(TRI_ERROR_QUERY_KILLED, /*sync*/ false);
+  if (ServerState::instance()->isCoordinator()) {
+    if (!_queryKilled.load(std::memory_order_acquire) &&
+        !_queryKilled.exchange(true, std::memory_order_acq_rel)) {
```

Comment:
No doubt this works, but I think this is a performance optimization that will only have an effect when a query is killed, which should happen very seldomly. At least I think the use case "killing a query" doesn't need to be performance-optimized.

Reply:
I think you are right and only exchange will be look better